### DISCUSSION TITLE
Reddit API changes

### DISCUSSION THREAD

**Discussion by Outlet2048** [Docile-Alligator/Infinity-For-Reddit#1437] [General]
Hi
Just read about this:
https://www.reddit.com/r/reddit/comments/12qwagm/an_update_regarding_reddits_api/

How will it affect our usage of this app?

**Comment by kaanyalova** [comment]
More details here https://www.reddit.com/r/apolloapp/comments/12ram0f/had_a_few_calls_with_reddit_today_about_the/

**Comment by ste5e99** [comment]
It seems like they want to move third-parties onto their free Data API plan, while rescinding certain types of content behind paywalled premium APIs, mostly NSFW content. u/KeyserSosa explains a bit to one of the WSB bot creators here: https://www.reddit.com/r/reddit/comments/12qwagm/an_update_regarding_reddits_api/jgruulz/

**Comment by Synicalx** [comment]
I'd be very interested to know if in future, apps (not necessarily just Inifinity, I mean any apps) will allow users to import their own API keys to use instead. The techcrunch article I saw on this topic seems to imply that there are some vaguely defined non-commercial uses of API keys that will still be offered to people for free. That is of course assuming it's all the same API/endpoints. 

"It’s not a blanket policy change. As reported by The New York Times, Reddit’s API will remain free to developers who want to build apps and bots that help people use Reddit, as well as to researchers who wish to study Reddit for strictly academic or noncommercial purposes."

Source: https://techcrunch.com/2023/04/18/reddit-will-begin-charging-for-access-to-its-api/

**Comment by Outlet2048** [comment]
Is there any difference in how this affects users that use the app with an account vs those thag use anonymous mode?

**Reply by 7eter** [reply]
> Is there any difference in how this affects users that use the app with an account vs those thag use anonymous mode?

as far as i understand probably both will be affected. Since in both cases the API is used - weather that's with or without authentication. So there will either be a fee one has to pay to reddit in order to use Infinity, and other 3rd party apps or someone decides to web-scrape reddit, or reddit decides to give away free keys for specific purposes

**Comment by mustafakamran98** [comment]
It's gotten worse https://www.reddit.com/r/apolloapp/comments/13ws4w3/had_a_call_with_reddit_to_discuss_pricing_bad/

**Reply by Outlet2048** [reply]
This is horrible, who's supposed to pay that

**Reply by IcedEagle** [reply]
That's absurd. Reddit is clearly pricing third-party apps out of the ecosystem. Without Infinity, I simply won't use Reddit on mobile. The official app is not a good experience.

**Comment by Gitchick57** [comment]
Looks like Reddit will crash and burn very soon.  



May 31, 2023, 4:19 PM by ***@***.***:

**Comment by ff2400t** [comment]
https://www.theverge.com/2023/5/31/23743993/reddit-apollo-client-api-cost

**Comment by mustafakamran98** [comment]
https://reddit.com/r/modnews/comments/13wshdp/api_update_continued_access_to_our_api_for/ Looking at this, would allowing users to provide their own API keys be a viable option? That way each user is responsible for their own amount of requests.

**Comment by DraconicNEO** [comment]
@Docile-Alligator have you ever considered using a web scraping based approach, similar to something like Newpipe does for YouTube? 

Just wondering since doing that could possibly open up a lot more options like Reddit chat features that aren't available in the API but also allowing this app to go forward if they decide to close the API down completely. 

Why should we use their shitty limited API just so they can lock us out later?

**Reply by shbhmnk** [reply]
Possibly for legal reasons

**Reply by DraconicNEO** [reply]
> Possibly for legal reasons

Web scraping is not illegal, if it was Newpipe wouldn't exist or would have been taken down long ago.

**Comment by rom4ster** [comment]
@Docile-Alligator  Reddit sessions NEVER expire. It should be likely that you can get a token or something from that and then use whatever calls the official mobile app makes? Then you just have to ask users to sign in (secure this) 

Edit: It sseems that you can just use the API lol just swap out the dev key with one generated by the user or something.

**Comment by ste5e99** [comment]
I don't expect any new progress on the Infinity front, but here's Apollo's lead developer announcing the closure of Apollo for good.

https://old.reddit.com/r/apolloapp/comments/144f6xm/apollo_will_close_down_on_june_30th_reddits/

**Reply by mustafakamran98** [reply]
https://www.theverge.com/2023/6/8/23754616/reddit-third-party-apps-api-shutdown-rif-reddplanet-sync more third party reddit apps are calling it quits

**Reply by rom4ster** [reply]
Im tired of this narrative. These closed source apps dont have a way out, but we do. Infinity is an OPEN SOURCE app so we can utilize that fact to our advantage. Even if the dev wont do it someone else can.

**Comment by Gitchick57** [comment]
And so it begins...
https://www.theverge.com/2023/6/8/23754183/apollo-reddit-app-shutting-down-api

I assume infinity will be next. 😢



Jun 7, 2023, 9:56 PM by ***@***.***:

**Reply by DraconicNEO** [reply]
Wish people would even acknowledge the idea of web scraping, it's not an alien or foreign concept. We've just been spoiled by Reddit giving us Unconditional API Access for the longest time.

**Reply by ste5e99** [reply]
> Wish people would even acknowledge the idea of web scraping, it's not an alien or foreign concept. We've just been spoiled by Reddit giving us Unconditional API Access for the longest time.

Feel free to get started and let us know how it goes. It's not as easy as you might think.

**Reply by DraconicNEO** [reply]
> > Wish people would even acknowledge the idea of web scraping, it's not an alien or foreign concept. We've just been spoiled by Reddit giving us Unconditional API Access for the longest time.
> 
> Feel free to get started and let us know how it goes. It's not as easy as you might think.

It might not be but if NewPipe can do it why shouldn't we be able to as well?
Or are you just trying to dismiss the idea by implying it's impossible or too difficult...

**Reply by Logics4** [reply]
There seems to be some work done on web scraping already, you may check out [Teddit](https://codeberg.org/teddit/teddit) - they even offer an API to use, it could be an alternative to the official one.

**Reply by rom4ster** [reply]
You guys would seriously consider WEB SCRAPING over simply using your own api keys/tokens? why?  This is not crapple, android sideloading is a piece of cake.

**Reply by DraconicNEO** [reply]
> You guys would seriously consider WEB SCRAPING over simply using your own api keys/tokens? why? This is not crapple, android sideloading is a piece of cake.

Web Scraping offers true freedom from Reddit's API policies, Personal API keys can be revoked or denied by the overlords at any moment, effectively shutting down that option completely. Web scraping can be broken by updates but unless Reddit shuts down their website entirely it will always be possible to access by scraping. After all Services like NewPipe do it, I bet if they used some YouTube API keys they would've had their plug pulled long ago (YouTube hates third party apps more than Reddit by a long shot).

It's not about Apple vs. Android (Also I don't think apple would be happy about web scraping for reasons above) it's about a more sustainable since Reddit is going to make these APIs which we've been oh so spoiled by harder to use due to certain restrictions if not impossible.

IMO I have the same opinion on continued API use that you do about web scraping because in the future when they start cutting people off because too many people are using the API it won't be viable anymore.

**Reply by shbhmnk** [reply]
> You guys would seriously consider WEB SCRAPING over simply using your own api keys/tokens? why?  This is not crapple, android sideloading is a piece of cake. 

Not sure if you are aware of the new rate changes for these keys. Reddit is essentially making keys inaccesibly by setting these rates.

**Reply by rom4ster** [reply]
> > You guys would seriously consider WEB SCRAPING over simply using your own api keys/tokens? why?  This is not crapple, android sideloading is a piece of cake.
> 
> Not sure if you are aware of the new rate changes for these keys. Reddit is essentially making keys inaccesibly by setting these rates.

100 calls a minute for oauth keys at free tier right?

**Reply by shbhmnk** [reply]
> > > You guys would seriously consider WEB SCRAPING over simply using your own api keys/tokens? why?  This is not crapple, android sideloading is a piece of cake.
> > 
> > Not sure if you are aware of the new rate changes for these keys. Reddit is essentially making keys inaccesibly by setting these rates.
> 
> 100 calls a minute for oauth keys at free tier right?
> 

An app of such scale as infinity cannot survive with 100 calls. See thier point of view as well https://www.reddit.com/r/apolloapp/comments/13ws4w3/had_a_call_with_reddit_to_discuss_pricing_bad/, https://www.reddit.com/r/redditsync/comments/142dzy8/the_future_of_sync/

If you think you can sustain the app with these rates then feel free to continue. The app is open source, give it a shot. It will help everyone if infinity continues.

**Reply by mustafakamran98** [reply]
> > > > You guys would seriously consider WEB SCRAPING over simply using your own api keys/tokens? why?  This is not crapple, android sideloading is a piece of cake.
> > > 
> > > 
> > > Not sure if you are aware of the new rate changes for these keys. Reddit is essentially making keys inaccesibly by setting these rates.
> > 
> > 
> > 100 calls a minute for oauth keys at free tier right?
> 
> An app of such scale as infinity cannot survive with 100 calls. See thier point of view as well https://www.reddit.com/r/apolloapp/comments/13ws4w3/had_a_call_with_reddit_to_discuss_pricing_bad/, https://www.reddit.com/r/redditsync/comments/142dzy8/the_future_of_sync/
> 
> If you think you can sustain the app with these rates then feel free to continue. The app is open source, give it a shot. It will help everyone if infinity continues.

Pretty sure they're referring to the scenario of the users using their own API keys rather than an app-wide API key like before.

**Reply by rom4ster** [reply]
> > > > You guys would seriously consider WEB SCRAPING over simply using your own api keys/tokens? why?  This is not crapple, android sideloading is a piece of cake.
> > > 
> > > 
> > > Not sure if you are aware of the new rate changes for these keys. Reddit is essentially making keys inaccesibly by setting these rates.
> > 
> > 
> > 100 calls a minute for oauth keys at free tier right?
> 
> An app of such scale as infinity cannot survive with 100 calls. See thier point of view as well https://www.reddit.com/r/apolloapp/comments/13ws4w3/had_a_call_with_reddit_to_discuss_pricing_bad/, https://www.reddit.com/r/redditsync/comments/142dzy8/the_future_of_sync/
> 
> If you think you can sustain the app with these rates then feel free to continue. The app is open source, give it a shot. It will help everyone if infinity continues.

Do you honestly think a single reddit user will pass 100 calls per minute? Your linked articles provide absolutely zero evidence of that. Please provide actual evidence of your claims instead of closed source apps who (obviously) cannot sustain that rate shutting down.

**Reply by Logics4** [reply]
> You guys would seriously consider WEB SCRAPING over simply using your own api keys/tokens? why? This is not crapple, android sideloading is a piece of cake.

The API changes also include removing all content tagged as NSFW from showing up in API calls, meaning, through the API, you cannot see any of it, except through the official app (which uses private endpoints). Something like Teddit's API (as I mentioned previously) could be an alternative, although it probably wouldn't allow you to login with your account but the option to use it could be given in addition to using your own keys.

**Reply by DraconicNEO** [reply]
> > Wish people would even acknowledge the idea of web scraping, it's not an alien or foreign concept. We've just been spoiled by Reddit giving us Unconditional API Access for the longest time.
> 
> Reddit has public RSS feeds. No need to scrape.
> 
> See https://www.reddit.com/wiki/rss/
> 
> How long that stays available is unknown. Twitter is hostile to RSS. I suspect Reddit will pull the RSS feed also.

Hence why we should invest in options that aren't directly controlled by them. We've been spoiled rotten by the API's easy availability, we shouldn't make the same mistake depending on something they could later just pull.

**Comment by snvoid** [comment]
Seems they are providing exceptions for non-commercial accessibility-focused apps. Maybe Infinity can use that angle and then use scraping for nsfw content or teddit's api. Someone also posted https://archive.ph/gcQVK to HN which might be useful

RedReader got an exception and will be able to continue existing with access to the free data apis 

![image](https://github.com/Docile-Alligator/Infinity-For-Reddit/assets/60519765/0da887ba-53f7-4a61-a7e5-46c7a6bdca56)

**Comment by snvoid** [comment]
Can also read some updates from Dystopia for Reddit. Apparently they were able to come to an agreement regarding access for free as well with similar considerations as RedReader. It seems that they can accept donations as well. 

https://www.reddit.com/r/DystopiaForReddit/comments/145e9sk/update_dystopia_will_continue_operating_for_free/ 

> Hi all!
> 
> After quite a bit of uncertainty and worry, I'm happy to finally share some good news with y'all: Dystopia is here to stay and will continue to be completely free! For those interested in the full story, read on.
> 
> For context, last week Reddit announced [some major changes to the limits and pricing of their API](https://reddit.com/r/modnews/comments/13wshdp/api_update_continued_access_to_our_api_for/). In short, these changes required most high-volume consumers of Reddit data, including third-party applications like Dystopia, to pay for access to Reddit through a new service which they've dubbed the "Data API". This posed a rather existential problem to Dystopia as this app both brings in precisely zero dollars in revenue and because I have zero plans, let alone means, to properly monetize the app. Of course, Dystopia is not alone in this struggle and indeed these changes have also been widely felt across the third-party app ecosystem. As I'm sure many of you have already heard, these changes have pushed numerous other apps which users depend on to be able to access Reddit---from Apollo, to Reddit is Fun, to Sync---to either shut down or, in the case of Narwhal, switch to a subscription only model.
> 
> On June 7th, however, Reddit reached out and scheduled a call with me to discuss how Dystopia can continue to operate. In this call, Reddit expressed its commitment to accessibility and a desire to build strong partnerships with accessibility focused apps so that they can better support the community. While I did (gently) poke at them over the fact that doing all of this now could be interpreted as just a reaction to bad PR, they quickly clarified to me that the impact on accessibility apps was wholly unintended and that they are working hard to rectify it. While you'll just have to take my word for it, I really did get the impression that this was sincere, especially given that Reddit could have easily ignored the issue and swept accessibility under the rug like so many other technology companies have.
> 
> In any case, Reddit has offered me an agreement which provides Dystopia with free access to the Data API so long as Dystopia remains unmonetized (though donations are permitted) and provided that it be principally focused on serving disabled, blind, and low vision users. Of note, these terms are largely non-binding for both Reddit and I, and so should I wish to monetize the app in the future (which I have no plans to do so) I may exit the agreement. Similarly, should Reddit wish to stop offering free API access to Dystopia, they too are free to exit. As these terms feel fair and match well with my plans for Dystopia, I was happy to accept this agreement and continue to offer Dystopia to everyone for free.
> 
> So, what's next for Dystopia? Well, in light of the closure of many other apps which people rely on to access Reddit, I've finally submitted Dystopia to the App Store for review. Due to both the technical complexity and other issues around submitting applications which can provide access to user generated content (and particularly adult content) to the App Store, I can't quite say for sure when it will be available but hopefully we'll all be off TestFlight very soon. The intention, however, is for it to be available on or before the end of this month to support users who need to transition off of other apps in order to keep using Reddit. It is unfortunate that Dystopia needs to finally officially launch under these circumstances but for better or for worse I hope it provides some continuity for those who are losing access to their favorite apps but still want to stick around on Reddit.
> 
> Additionally, due to API rule changes, Dystopia still will likely not support NSFW content come July 1st. I'll see what I can do to paper over this as I do believe that everyone should be able to access all of Reddit, if they so choose. There are some legal and technical issues that I need to more carefully consider, but stay tuned.
> 
> Thank you all for your continued support, it's always appreciated ❤️

**Comment by ixfd64** [comment]
Another idea is to use the app secrets extracted from the official app. See: https://news.ycombinator.com/item?id=36086240

**Reply by DraconicNEO** [reply]
While that could definitely work it could cause legal issues if a project were to use it. 
Could be an idea to use it to make apps not hosted on Github though (there are other platforms that aren't DMCA friendly where you can host apps like that).

**Reply by rom4ster** [reply]
Lol , you just need to have the functionality and have users input the secrets and github will be cool with it.

**Reply by rom4ster** [reply]
LOL it just got deleted, nothing dies on the internet


   func (c *Client) generateResultData(epoch int64) string {
        return fmt.Sprintf("Epoch:%d|User-Agent:%s|Client-Vendor-ID:%s", epoch, c.userAgent, c.clientVendorID)
    }

    func (c *Client) generateBodyData(epoch int64, body []byte) string {
        return fmt.Sprintf("Epoch:%d|Body:%s", epoch, string(body))
    }

    func hmacSignData(data string) string {
        digester := hmac.New(sha256.New, []byte(REDDIT_ANDROID_HMAC_SECRET))
        digester.Write([]byte(data))
        return hex.EncodeToString(digester.Sum(nil))
    }

    REDDIT_ANDROID_OAUTH_CLIENT_ID = "ohXpoqrZYub1kg"
    REDDIT_ANDROID_HMAC_SECRET = "8c7abaa5f905f70400c81bf3a1a101e75f7210104b1991f0cd5240aa80c4d99d"

**Reply by rom4ster** [reply]
full page 



It would be a real shame if libreddit were to log the requests made by the Android app and try to replicate them, using this code to sign the "x-hmac-signed-body" and "x-hmac-signed-result" headers in the [https://accounts.reddit.com/api/login](https://web.archive.org/web/20230601193626/https://accounts.reddit.com/api/login) request.
    func (c *Client) generateResultData(epoch int64) string {
        return fmt.Sprintf("Epoch:%d|User-Agent:%s|Client-Vendor-ID:%s", epoch, c.userAgent, c.clientVendorID)
    }

    func (c *Client) generateBodyData(epoch int64, body []byte) string {
        return fmt.Sprintf("Epoch:%d|Body:%s", epoch, string(body))
    }

    func hmacSignData(data string) string {
        digester := hmac.New(sha256.New, []byte(REDDIT_ANDROID_HMAC_SECRET))
        digester.Write([]byte(data))
        return hex.EncodeToString(digester.Sum(nil))
    }

    REDDIT_ANDROID_OAUTH_CLIENT_ID = "ohXpoqrZYub1kg"
    REDDIT_ANDROID_HMAC_SECRET = "8c7abaa5f905f70400c81bf3a1a101e75f7210104b1991f0cd5240aa80c4d99d"
And then use:
    https://oauth.reddit.com/r/%s/about.json
    https://oauth.reddit.com/r/%s/top.json
    https://oauth.reddit.com/search.json
etc.

**Reply by ixfd64** [reply]
> LOL it just got deleted, nothing dies on the internet

It's still up for me.

**Reply by rom4ster** [reply]
came back up.

**Reply by DraconicNEO** [reply]
I wonder if Reddit could potentially do anything to counter this on a technical side? 
i.e. checking for differences from regular app in function and suspending accounts for "suspicious activity",
Would probably hurt them much more than the API thing itself at this point though...

**Reply by darkevilmac** [reply]
@DraconicNEO it's probably possible, it just depends how much they really care on their end. Odds are a lot of this is to make investors see "Hey look at how many people are using our app!" - if a third party client effectively spoofs the official one then their metrics look good.

**Reply by rom4ster** [reply]
It seems difficult to do so without some technical work, and we already know from their app design how wonderfully they do technical work /s. if you do exactly what the app is doing, then it would be hard to know that it came from somewhere else. To them a user using the normal api for normal things, and then using this private one for nsfw, would look like a user using a 3rd party app and then switching to the original for NSFW. As it is currently designed by my interpretation of it.

**Reply by darkevilmac** [reply]
From my cursory glance at the code through a disassembler it honestly looks like it's just a bunch of GraphQL operations alongside a normal REST endpoint for the initial login flow(s). Their login token is just a hardcoded string in the APK in an XML file that gets encoded to base64 at runtime.

**Reply by rom4ster** [reply]
oh in that case, if its in the XML file, why not have infinity just STEAL it from their XML file. its just a folder so easy enough.

**Comment by darkevilmac** [comment]
Been experimenting a bit just to see what sort of info can be extracted from the APK.

Trying to make this pretty agnostic towards any obfuscation in the future which is why I'm not relying on class names.

https://github.com/darkevilmac/FedUp
